adapter/avro/src/main/java/org/apache/arrow/adapter/avro/producers/AvroNullableProducer.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.arrow.adapter.avro.producers;
+
+import java.io.IOException;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.avro.io.Encoder;
+
+/**
+ * Producer wrapper which producers nullable types to an avro encoder. Write the data to the
+ * underlying {@link FieldVector}.
+ */
+public class AvroNullableProducer<T extends FieldVector> extends BaseAvroProducer<T> {
+
+  private final Producer<T> delegate;
+
+  /** Instantiate a AvroNullableProducer. */
+  public AvroNullableProducer(Producer<T> delegate) {
+    super(delegate.getVector());
+    this.delegate = delegate;
+  }
+
+  @Override
+  public void produce(Encoder encoder) throws IOException {
+    if (vector.isNull(currentIndex)) {
+      encoder.writeInt(1);
```

Comment:
Does it mean that you placed `[type, null]` as the union type? Is it a good convention to put `null` before `type` so the default value is null when reader type does not match the writer type?

Reply:
I tried doing that, it broke the consumers and the data would not do round trip. In Avro this is considered as the difference between "nullable" and "optional", with "optional" being the semantics you are suggesting. I have kept it as nullable for now. I do think we will need to follow up with a PR on the consumers, particularly to address null handling. Nullable types get mangled in the schema round trip at the moment!